c10/impl/DeviceGuardImplInterface.h

```diff
@@ -80,6 +80,11 @@ struct C10_API DeviceGuardImplInterface {
    */
   virtual Stream exchangeStream(Stream) const noexcept = 0;
 
+  /**
+   * Get the number of devices.
+   */
+  virtual DeviceIndex deviceCount() const = 0;
```

Comment:
This style of interface always strikes me as strange because the returned type is an Index but the value returned is definitely _not_ a valid index.  But since @ezyang told you to do this, I'll approve.